dsa: Add e2e tests

ozhuraki opened the pull request:
Closes: #1026 
Closes: #642

bart0sh commented on test/e2e/dsa/dsa.go:
```diff
@@ -0,0 +1,65 @@
+			framework.DumpAllNamespaceInfo(f.ClientSet, f.Namespace.Name)
+			kubectl.LogFailedContainers(f.ClientSet, f.Namespace.Name, framework.Logf)
+			framework.Failf("unable to wait for all pods to be running and ready: %v", err)
+		}
```
Here we should at least check if DSA resource(s) are available on the node.

mythi replied:
This is still missing. For others we have, e.g.,
https://github.com/intel/intel-device-plugins-for-kubernetes/blob/1e6da3001f6989c2b14e97286f7668aae604a51f/test/e2e/qat/qatplugin_dpdk.go#L77-L80

bart0sh reviewed (changes requested):
Check if plugin announces DSA resources properly

mythi commented:
> Check if plugin announces DSA resources properly

Also, prepare with changes required from #985 (i.e., pod security admission)

bart0sh commented:
> > Check if plugin announces DSA resources properly
> 
> Also, prepare with changes required from #985 (i.e., pod security admission)

And add running[ test workload](https://github.com/intel/intel-device-plugins-for-kubernetes/tree/main/demo/dsa-accel-config-demo).

ozhuraki commented:
@bart0sh 
> Here we should at least check if DSA resource(s) are available on the node.

This is checked now, the demo scheduling depends on the DSA resources.

@mythi 
> Also, [...] (i.e., pod security admission

Done

@bart0sh @mythi 

Updated, take a look please

mythi commented on demo/dsa-accel-config-demo-pod.yaml:
```diff
@@ -2,15 +2,36 @@ apiVersion: v1
-    image: dsa-accel-config-demo:devel
+    image: intel/dsa-accel-config-demo:devel
+    command: [ "/bin/bash", "-e", "-c", "cd /test; ./dsa_test -v -w 0 -d dsa0/wq0.0 -f 0x0 -o 0x3; while true; do sleep 300000; done" ]
+    securityContext:
+      privileged: true
```
some debug leftover?

ozhuraki replied:
Actually no, seems like this is required at least for the bmetal. It's unclear to me if this ever worked (bisecting, cross-checking)

mythi commented on demo/dsa-accel-config-demo/idxd-reset.patch:
```diff
@@ -0,0 +1,26 @@
+From 07e1137a71021dc02df98b942a260b70ff852ca5 Mon Sep 17 00:00:00 2001
+From: Oleg Zhurakivskyy <oleg.zhurakivskyy@intel.com>
+Date: Tue, 3 May 2022 15:28:05 +0300
+Subject: [PATCH] idxd: Ignore reset device error
```
it's not clear why this is needed. have we reported this bug to `idxd-config` upstream?

mythi commented on deployments/dsa_plugin/overlays/dsa_initcontainer/dsa_initcontainer.yaml:
```diff
@@ -17,13 +17,18 @@ spec:
           privileged: true
         volumeMounts:
+        - mountPath: /sys/bus/dsa
+          name: sys-bus-dsa
```
do we need it for the operator too?

mythi reviewed (commented):
it looks this PR needs to be split into three separate PRs: 1) YAML fixes, 2) `test/e2e/dsa` (this PR), and 3) github workflows part

mythi commented on demo/dsa-accel-config-demo/Dockerfile:
```diff
@@ -1,9 +1,10 @@
-RUN dnf install -y libuuid-devel json-c-devel kmod-devel libudev-devel
+RUN dnf install -y \
+    wget autoconf automake libtool m4 diffutils file make dnf-utils patch \
+    libuuid-devel json-c-devel kmod-devel libudev-devel
```
was it so that `kmod` and `udev` where not needed?
```suggestion
    libuuid-devel json-c-devel
```

ozhuraki replied:
Yes, both not needed, dropped them

bart0sh reviewed (dismissed):
/lgtm

@mythi please, don't merge yet. We need to test it first.

ozhuraki commented:
@bart0sh @mythi 
Changes as discussed, please take a look:
- Added idxd reset patch to the dsa demo
- Added an overlay for the dsa demo for the bmetal environment

mythi commented:
note: `dsa-accel-config-demo` is not yet built by CI because the job gets cancelled early.

bart0sh commented on demo/dsa-accel-config-demo/idxd-reset.patch:
```diff
@@ -0,0 +1,26 @@
+ 	rc = accfg_set_param(ctx, dfd, "cmd_status", "1", 1);
+ 	/* older drivers don't support writing to cmd_status */
+-	if (rc && rc != -EACCES) {
++	if (0 && rc && rc != -EACCES) {
```
Does this mean we ignore accfg_set_param return code?

bart0sh commented on test/e2e/dsa/dsa.go:
```diff
@@ -0,0 +1,83 @@
+		framework.Failf("unable to locate %q: %v", demoYaml, err)
+	}
+
+	ginkgo.It("checks availability of DSA resources", func() {
```
```suggestion
        ginkgo.It("runs DSA plugin and a demo workload", func() {
```

bart0sh commented on test/e2e/dsa/dsa.go:
```diff
@@ -0,0 +1,83 @@
+		framework.RunKubectlOrDie(f.Namespace.Name, "--namespace", f.Namespace.Name, "apply", "-k", filepath.Dir(demoPath))
+
+		ginkgo.By("waiting for the DSA demo to succeed")
+		f.PodClient().WaitForSuccess("dsa-accel-config-demo", 300*time.Second)
```
Can we add something like this here? https://github.com/intel/intel-device-plugins-for-kubernetes/blob/main/test/e2e/gpu/gpu.go#L97 to see the test output or its tail if it's too long?

mythi reviewed (approved):
LGTM, thanks!

bart0sh commented:
/lgtm

@ozhuraki can we add .github/workflows/e2e-dsa.yml to this PR to see if dsa test passes?

ozhuraki commented:
@mythi @bart0sh 
Changes, please take a look:
- Dropped privileged for the demo
- Dropped customization overlay for bmetal CI
- Added a log output for the demo
- Refined idxd-reset.patch (needed for kernels 5.17.x)
- Extended the demo coverage

mythi commented:
> @ozhuraki can we add .github/workflows/e2e-dsa.yml to this PR to see if dsa test passes?

I asked to keep that separate so that the two commits are split.

bart0sh commented:
@mythi how can we see if the DSA CI job working then?

mythi commented:
> @mythi how can we see if the DSA CI job working then?

let's then cherry-pick the commit from #1026 and have two commits here. Normally the e2e test case development (e.g. this PR) has been independent from setting up the CI (#1026) but this time it's a rather special case.

ozhuraki commented:
Changes:
- Added a CI workflow for e2e DSA
